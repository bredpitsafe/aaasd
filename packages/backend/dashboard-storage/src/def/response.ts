/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

/**
 * @deprecated
 */
export type DashboardItem = DashboardsListItem & {
  draftDigest: string;
};


// TODO: remove this file after enums transfer into more appropriate place and tests fix
/**
 * @deprecated
 */
export interface Response {
  error?: ErrorResponse;
  payload?:
    | UnsubscribeResponse
    | PingResponse
    | HeartbeatResponse
    | AuthenticateResponse
    | LogoutResponse
    | SubscribeToDashboardResponse
    | SubscribeToDashboardsListResponse
    | FetchDashboardConfigResponse
    | FetchDashboardDraftResponse
    | CreateDashboardResponse
    | ImportDashboardResponse
    | UpdateDashboardResponse
    | UpdateDashboardDraftResponse
    | ResetDashboardDraftResponse
    | UpdateDashboardPermissionsResponse
    | UpdateDashboardShareSettingsResponse
    | DeleteDashboardResponse
    | SubscribeToDashboardPermissionsResponse
    | FetchDashboardIdByLegacyIdResponse
    | RenameDashboardResponse
    | UpdateDashboardScopeBindingResponse;
  timestamp: string;
  traceId: string;
  correlationId: number;
  state: ResponseState;
}
/**
 * @deprecated
 */
export interface ErrorResponse {
  kind: string;
  description: string;
  args?: {
    [k: string]: unknown;
  };
}
/**
 * @deprecated
 */
export interface UnsubscribeResponse {
  type: "Unsubscribed";
}
/**
 * @deprecated
 */
export interface PingResponse {
  type: "Pong";
}
/**
 * @deprecated
 */
export interface HeartbeatResponse {
  type: "Heartbeat";
}
/**
 * @deprecated
 */
export interface AuthenticateResponse {
  type: "Authenticated";
  username: string;
}
/**
 * @deprecated
 */
export interface LogoutResponse {
  type: "LoggedOut";
}
/**
 * @deprecated
 */
export interface SubscribeToDashboardResponse {
  type: "Dashboard";
  dashboard: DashboardItem;
}
/**
 * @deprecated
 */
export interface DashboardsListItem {
  id: string;
  name: string;
  kind: Kind;
  status: Status;
  digest?: string;
    // This is update time actually
  insertionTime: string;
  hasDraft: boolean;
  sharePermission: SharePermission;
  permission: Permission;
  owners: User[];
  permissionsCount?: number;
  scopes: Scope[];
}
/**
 * @deprecated
 */
export interface User {
  user: string;
}
/**
 * @deprecated
 */
export interface Scope {
  [k: string]: string | number | boolean | null;
}
/**
 * @deprecated
 */
export interface SubscribeToDashboardsListResponse {
  type: "DashboardsList";
  list: DashboardsListItem[];
}
/**
 * @deprecated
 */
export interface FetchDashboardConfigResponse {
  type: "DashboardConfig";
  config: string;
  digest: string;
}
/**
 * @deprecated
 */
export interface FetchDashboardDraftResponse {
  type: "DashboardDraft";
  draft: string;
  digest: string;
}
/**
 * @deprecated
 */
export interface CreateDashboardResponse {
  type: "DashboardCreated";
  id: string;
}
/**
 * @deprecated
 */
export interface ImportDashboardResponse {
  type: "DashboardImported";
  id: string;
}
/**
 * @deprecated
 */
export interface UpdateDashboardResponse {
  type: "DashboardUpdated";
}
/**
 * @deprecated
 */
export interface UpdateDashboardDraftResponse {
  type: "DashboardDraftUpdated";
}
/**
 * @deprecated
 */
export interface ResetDashboardDraftResponse {
  type: "DashboardDraftReset";
}
/**
 * @deprecated
 */
export interface UpdateDashboardPermissionsResponse {
  type: "PermissionsUpdated";
}
/**
 * @deprecated
 */
export interface UpdateDashboardShareSettingsResponse {
  type: "DashboardShareSettingsUpdated";
}
/**
 * @deprecated
 */
export interface DeleteDashboardResponse {
  type: "DashboardDeleted";
}
/**
 * @deprecated
 */
export interface SubscribeToDashboardPermissionsResponse {
  type: "DashboardPermissionsList";
  list: {
    user: string;
    permission: string;
  }[];
}
/**
 * @deprecated
 */
export interface FetchDashboardIdByLegacyIdResponse {
  type: "DashboardId";
  id: string;
}
/**
 * @deprecated
 */
export interface RenameDashboardResponse {
  type: "DashboardRenamed";
}
/**
 * @deprecated
 */
export interface UpdateDashboardScopeBindingResponse {
  type: "DashboardScopeBindingUpdated";
  platformTime: string;
}

// TODO: move this enums to the approriate place
export const enum Kind {
  User = "User",
  Robot = "Robot"
}
export const enum Status {
  Active = "Active",
  Suspended = "Suspended",
  Archived = "Archived"
}
export const enum SharePermission {
  None = "None",
  Viewer = "Viewer",
  Editor = "Editor"
}
export const enum Permission {
  None = "None",
  Viewer = "Viewer",
  Editor = "Editor",
  Owner = "Owner"
}
/**
 * @deprecated
 */
export const enum ResponseState {
  Done = "Done",
  InProgress = "InProgress"
}
