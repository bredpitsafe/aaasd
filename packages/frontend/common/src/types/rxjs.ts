import type { Observable, Subject, Subscription } from 'rxjs';

export interface IObserver<T, E> {
    next: (value: T) => void;
    error: (err: E) => void;
    complete: () => void;
}

export interface IObservable<T, E> extends Observable<T> {
    subscribe(next: (value: T) => void): Subscription;
    subscribe(observer?: Partial<IObserver<T, E>>): Subscription;
}

export interface ISubject<T, E> extends Subject<T> {
    error(err: E): void;
    subscribe(next: (value: T) => void): Subscription;
    subscribe(observer?: Partial<IObserver<T, E>>): Subscription;
}
