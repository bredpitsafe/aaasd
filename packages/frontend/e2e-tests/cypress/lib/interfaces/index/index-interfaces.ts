export enum EPagesName {
    dashboard = 'Dashboard',
    tradingServersManager = 'Trading Servers Manager',
    herodotusTerminal = 'Herodotus Terminal',
    herodotusTrades = 'Herodotus Trades',
    herodotusDashboard = 'Herodotus Dashboard',
    tradingStats = 'Trading Stats',
    backtestingManager = 'Backtesting Manager',
    jsonCharts = 'JSON Charts',
    balanceMonitor = 'Balance Monitor',
    wsQueryTerminal = 'WS Query Terminal',
}
