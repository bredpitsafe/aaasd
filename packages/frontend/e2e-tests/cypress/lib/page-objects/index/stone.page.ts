import { BasePage } from '../../base.page';
import { Button } from '../../base/elements/button';
import { Text } from '../../base/elements/text';
enum EStonePageSelectors {
    Dashboard = '[href*=dashboard]',
    TradingServersManager = '[href*=trading-servers-manager]',
    HerodotusTerminal = '[href*=herodotus-terminal]',
    TradingStats = '[href*=trading-stats]',
    BacktestingManager = '[href*=backtesting]',
    BalanceMonitor = '[href*=balance-monitor]',
    WSQueryTerminal = '[href*=ws-query-terminal]',
    JsonCharts = '[href*=charter-json-viewer]',
    LoadingIcon = '[class*="anticon anticon-loading"]',
}

class StonePage extends BasePage {
    readonly dashboard = new Text(EStonePageSelectors.Dashboard, false);
    readonly tradingServersManager = new Text(EStonePageSelectors.TradingServersManager, false);
    readonly herodotusTerminal = new Text(EStonePageSelectors.HerodotusTerminal, false);
    readonly tradingStats = new Text(EStonePageSelectors.TradingStats, false);
    readonly backtestingManager = new Text(EStonePageSelectors.BacktestingManager, false);
    readonly balanceMonitor = new Text(EStonePageSelectors.BalanceMonitor, false);
    readonly wsQueryTerminal = new Text(EStonePageSelectors.WSQueryTerminal, false);
    readonly jsonCharts = new Text(EStonePageSelectors.JsonCharts, false);
    readonly loadingIcon = new Button(EStonePageSelectors.LoadingIcon, false);

    checkElementsExists(): void {
        this.dashboard.checkExists();
        this.tradingServersManager.checkExists();
        this.herodotusTerminal.checkExists();
        this.tradingStats.checkExists();
        this.backtestingManager.checkExists();
        this.backtestingManager.checkExists();
    }

    open(): StonePage {
        this.visit();
        return this;
    }
}

export const stonePage = new StonePage();
