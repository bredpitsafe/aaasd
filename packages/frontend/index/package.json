{
  "name": "@frontend/index",
  "version": "1.0.0",
  "description": "simple index page for choose application",
  "scripts": {
    "build": "NODE_ENV=production npx webpack",
    "lint": "eslint './src/**/*.{js,ts,tsx}' --quiet",
    "server": "npx webpack serve",
    "test_unit": "jest --passWithNoTests",
    "types": "tsc --noEmit"
  },
  "dependencies": {
    "@frontend/common": "^1.0.1",
    "@common/types": "^1.0.0"
  }
}
