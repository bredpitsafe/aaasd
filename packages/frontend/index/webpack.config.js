const {
    buildConfig,
    htmlConfig,
    defaultInOut,
    defaultDevServerConfig,
} = require('../../../webpack/configs');
const path = require('path');

const app = 'index';
const resolvePath = () => path.resolve(__dirname);

module.exports = buildConfig([
    defaultInOut([app], resolvePath),
    htmlConfig([app], resolvePath),
    defaultDevServerConfig(app),
]);
