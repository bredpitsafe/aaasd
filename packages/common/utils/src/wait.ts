import { firstValueFrom, interval, map } from 'rxjs';

export const wait = (ms = 100): Promise<void> =>
    firstValueFrom<void>(interval(ms).pipe(map(() => undefined)));
