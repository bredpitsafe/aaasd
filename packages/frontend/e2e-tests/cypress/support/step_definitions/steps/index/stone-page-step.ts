import { Given } from '@badeball/cypress-cucumber-preprocessor';

import { EPagesName } from '../../../../lib/interfaces/index/index-interfaces';
import { ERobots } from '../../../../lib/interfaces/trading-servers-manager/robots-interfaces';
import { backtestingPage } from '../../../../lib/page-objects/backtesting/backtesting.page';
import { balanceMonitorPage } from '../../../../lib/page-objects/balance-monitor/balance-monitor.page';
import { charterJsonViewerPage } from '../../../../lib/page-objects/charter-json-viewer/charter-json-viewer.page';
import { authenticationPage } from '../../../../lib/page-objects/common/authentication.page';
import { settingsModal } from '../../../../lib/page-objects/common/settings.modal';
import { tableRow } from '../../../../lib/page-objects/common/table/table.row';
import { dashboardPage } from '../../../../lib/page-objects/dashboard/dashboard.page';
import { herodotusTerminalPage } from '../../../../lib/page-objects/herodotus-terminal/herodotus-terminal.page';
import { herodotusTradesPage } from '../../../../lib/page-objects/herodotus-trades/herodotus-trades.page';
import { stonePage } from '../../../../lib/page-objects/index/stone.page';
import { tradingServersManagerPage } from '../../../../lib/page-objects/trading-servers-manager/trading-servers-manager.page';
import { tradingStatsPage } from '../../../../lib/page-objects/trading-stats/trading-stats.page';
import { wsQueryTerminalPage } from '../../../../lib/page-objects/ws-query-terminal/ws-query-terminal.page';
import { customWait } from '../../../../lib/web-socket/server';
import { checkUrlInclude } from '../../../asserts/comon-url-assert';

Given(`user goes to the "Stone" page`, () => {
    stonePage.visit();
    stonePage.checkElementsExists();
});

Given(`user click an the {string} link`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.dashboard:
            stonePage.dashboard.click();
            break;
        case EPagesName.tradingServersManager:
            stonePage.tradingServersManager.click();
            break;
        case EPagesName.herodotusTerminal:
            stonePage.herodotusTerminal.click();
            break;
        case EPagesName.tradingStats:
            stonePage.tradingStats.click();
            break;
        case EPagesName.backtestingManager:
            stonePage.backtestingManager.click();
            break;
        case EPagesName.jsonCharts:
            stonePage.jsonCharts.click();
            break;
        case EPagesName.balanceMonitor:
            stonePage.balanceMonitor.click();
            break;
        case EPagesName.wsQueryTerminal:
            stonePage.wsQueryTerminal.click();
            break;
    }
});

Given(
    `user on the {string} page with the "default" backend server parameter`,
    (pageName: EPagesName) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.checkElementsExists();
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.checkElementsExists();
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.checkElementsExists();
                break;
            case EPagesName.backtestingManager:
                backtestingPage.checkElementsExists();
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.checkElementsExists();
                break;
        }
    },
);

Given(`user not sees data on the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.checkNotVisibleTabPanel();
            tradingServersManagerPage.checkVisibleLoading();
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.checkVisibleTabPanel();
            herodotusTerminalPage.checkTaskTableVisibility(false);
            stonePage.loadingIcon.checkVisible();
            break;
        case EPagesName.backtestingManager:
            backtestingPage.checkNotExistsTask();
            break;
    }
});

Given(`user sees the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.checkVisibleTabPanel();
            checkUrlInclude(tradingServersManagerPage.pageUrl);
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.checkVisibleTabPanel();
            herodotusTerminalPage.checkTaskTableVisibility(true);
            checkUrlInclude(herodotusTerminalPage.pageUrl);
            break;
        case EPagesName.tradingStats:
            tradingStatsPage.checkDailyVisibleTable();
            checkUrlInclude(tradingStatsPage.pageUrl);
            break;
        case EPagesName.backtestingManager:
            backtestingPage.checkVisiblePanel();
            checkUrlInclude(backtestingPage.pageUrl);
            break;
        case EPagesName.balanceMonitor:
            balanceMonitorPage.checkVisiblePanel();
            checkUrlInclude(balanceMonitorPage.pageUrl);
            break;
        case EPagesName.dashboard:
            dashboardPage.checkElementsExists();
            checkUrlInclude(dashboardPage.pageUrl);
            break;
        case EPagesName.jsonCharts:
            charterJsonViewerPage.checkElementsExists();
            checkUrlInclude(charterJsonViewerPage.pageUrl);
            break;
        case EPagesName.herodotusTrades:
            herodotusTradesPage.checkElementsExists();
            herodotusTradesPage.checkVisibleTable();
            checkUrlInclude(herodotusTradesPage.pageUrl);
            break;
        case EPagesName.wsQueryTerminal:
            wsQueryTerminalPage.checkElementsExists();
            wsQueryTerminalPage.checkVisibleTab();
            checkUrlInclude(wsQueryTerminalPage.pageUrl);
            break;
    }
});

Given(
    `user goes to the {string} page with the backend server parameter`,
    (pageName: EPagesName) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.openPageWithBackendServer();
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.openPageWithBackendServer();
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.openPageWithBackendServer();
                break;
            case EPagesName.backtestingManager:
                backtestingPage.openPageWithBackendServer();
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.openPageWithBackendServer();
                break;
            case EPagesName.wsQueryTerminal:
                wsQueryTerminalPage.openPageWithBackendServer();
                break;
        }
    },
);

Given(
    `user goes to the {string} page by {string} server params`,
    (pageName: EPagesName, nameServer: string) => {
        switch (pageName) {
            case EPagesName.tradingServersManager:
                tradingServersManagerPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.herodotusTerminal:
                herodotusTerminalPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.tradingStats:
                tradingStatsPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.backtestingManager:
                backtestingPage.openPageByBackendServer(nameServer);
                break;
            case EPagesName.balanceMonitor:
                balanceMonitorPage.openPageByBackendServer(nameServer);
                break;
        }
    },
);

Given(
    `user goes to the "Backtesting Manager" page with the backend server {string}`,
    (nameServer: string) => {
        backtestingPage.openPageWithBackendServerByParams(nameServer);
    },
);

Given(`user goes to the "Herodotus Terminal" page by index {string}`, (index: string) => {
    herodotusTerminalPage.openPageWithBackendServerByIndex(index);
});

Given(`user goes to the "Herodotus Terminal" page by name {string}`, (robotName: string) => {
    const index = `${ERobots[robotName]}`;
    herodotusTerminalPage.openPageWithBackendServerByIndex(index);
});

Given(`user click "Log Out" button on the {string} page`, () => {
    authenticationPage.clickLogOutButton();
});

Given(`user on the {string} page with the "ModalSettings" modal`, (pageName: EPagesName) => {
    settingsModal.checkElementsExists();
    customWait(1);
    switch (pageName) {
        case EPagesName.tradingServersManager:
            checkUrlInclude(tradingServersManagerPage.pageUrl);
            break;
        case EPagesName.herodotusTerminal:
            checkUrlInclude(herodotusTerminalPage.pageUrl);
            break;
        case EPagesName.tradingStats:
            checkUrlInclude(tradingStatsPage.pageUrl);
            break;
        case EPagesName.backtestingManager:
            checkUrlInclude(backtestingPage.pageUrl);
            break;
    }
});

Given(`user opens the {string} page`, (pageName: EPagesName) => {
    switch (pageName) {
        case EPagesName.dashboard:
            dashboardPage.visit();
            dashboardPage.checkElementsExists();
            break;
        case EPagesName.tradingServersManager:
            tradingServersManagerPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.herodotusTerminal:
            herodotusTerminalPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.tradingStats:
            tradingStatsPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.backtestingManager:
            backtestingPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.balanceMonitor:
            balanceMonitorPage.visit();
            settingsModal.checkElementsExists();
            break;
        case EPagesName.wsQueryTerminal:
            wsQueryTerminalPage.visit();
            settingsModal.checkElementsExists();
            break;
    }
});

Given(`user on the "Herodotus Terminal" page of the {string} robot`, (robotName: ERobots) => {
    checkUrlInclude(herodotusTerminalPage.pageUrl);
    herodotusTerminalPage.checkElementsExists();
    checkUrlInclude(`${ERobots[robotName]}`);
});

Given(
    `user types the modal settings with the {string} backend server parameter`,
    (backendServerName: string) => {
        if (backendServerName === 'default') {
            backendServerName = Cypress.env('backendServerName');
        }
        settingsModal.serverSelect.checkVisible();
        customWait(0.5);
        settingsModal.serverSelect.typeAndClickByText(backendServerName);
    },
);

Given(`user gets URL the "Charts" page and opens the page`, () => {
    tableRow.visitPageByLinkInRow();
});
