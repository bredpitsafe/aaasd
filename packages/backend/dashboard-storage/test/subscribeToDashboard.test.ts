import { wait } from '@common/utils/src/wait.js';
import type { TDashboardItem } from '@grpc-schemas/dashboard_storage-api-sdk/services/dashboard_storage/v1/common';
import { randomUUID } from 'crypto';
import { pick } from 'lodash-es';
import { firstValueFrom, share, shareReplay } from 'rxjs';
import { afterAll, beforeEach, describe, expect, it, vi } from 'vitest';

import { SharePermission } from '../src/def/response.ts';
import { generateCorrelationId } from './client/authentication.js';
import { Client, users } from './client/index.ts';
import initialDashboard from './fixtures/dashboard.json';

describe('SubscribeToDashboard', () => {
    const client = new Client({ authenticate: true });
    const otherClient = new Client({ authenticate: true, as: users.user2.username });
    let id = '';

    beforeEach(async () => {
        const createDashboardResponse = await client.request(initialDashboard);
        expect(createDashboardResponse.state).toBe('Done');
        expect(createDashboardResponse.payload.id).toMatch(/[a-f0-9-]+/);
        id = createDashboardResponse.payload.id;
    });

    afterAll(() => {
        client.close();
        otherClient.close();
    });

    it('gets initial state', async () => {
        const response = await client.request({ type: 'SubscribeToDashboard', id });
        expect(response.state).toBe('InProgress');
        expect(response.payload.dashboard).toMatchObject({
            id,
            ...pick(initialDashboard, 'name', 'kind', 'status'),
            hasDraft: false,
            permission: 'Owner',
        });
    });

    it('gets error when subscribing to non-existent dashboard', async () => {
        const response = await client.request({ type: 'SubscribeToDashboard', id: randomUUID() });

        expect(response.error).toMatchSnapshot();
    });

    it('gets error when subscribing to other user`s dashboard', async () => {
        const response = await otherClient.request({ type: 'SubscribeToDashboard', id });

        expect(response.error).toMatchSnapshot();
    });

    it('receives update after UpdateDashboard', async () => {
        const dashboard$ = client.send({ type: 'SubscribeToDashboard', id }).pipe(shareReplay(1));
        const subscriber = vi.fn();
        const subscription = dashboard$.subscribe(subscriber);
        // @ts-ignore TODO: remove this ts-ignore after tests migration to grpc
        const { dashboard } = (await firstValueFrom(dashboard$)).payload as {
            dashboard: TDashboardItem;
        };

        const updatedDashboard = {
            ...pick(dashboard, 'id', 'name', 'status', 'digest'),
            type: 'UpdateDashboard',
            config: 'updated config',
        };
        const updateResponse = await client.request(updatedDashboard);
        expect(updateResponse.state).toBe('Done');
        expect(updateResponse.payload.type).toBe('DashboardUpdated');

        await wait();
        expect(subscriber.mock.calls).toHaveLength(2);
        const updateDashboardResponse = subscriber.mock.calls[1][0];
        expect(updateDashboardResponse.state).toBe('InProgress');
        expect(updateDashboardResponse.payload.dashboard).toMatchObject({
            ...pick(dashboard, 'id', 'name', 'kind', 'status', 'hasDraft', 'permission'),
        });

        subscription.unsubscribe();
    });

    it('receives update after draft change', async () => {
        const dashboard$ = client.send({ type: 'SubscribeToDashboard', id }).pipe(shareReplay(1));
        const subscriber = vi.fn();
        const subscription = dashboard$.subscribe(subscriber);
        // @ts-ignore TODO: remove this ts-ignore after tests migration to grpc
        const { dashboard } = (await firstValueFrom(dashboard$)).payload as {
            dashboard: TDashboardItem;
        };

        const draft = {
            ...pick(dashboard, 'id'),
            type: 'UpdateDashboardDraft',
            config: 'config draft',
        };
        const updateDraftResponse = await client.request(draft);
        expect(updateDraftResponse.state).toBe('Done');
        expect(updateDraftResponse.payload.type).toBe('DashboardDraftUpdated');

        await wait();

        expect(subscriber.mock.calls).toHaveLength(2);
        const updateDashboardResponse = subscriber.mock.calls[1][0];
        expect(updateDashboardResponse.state).toBe('InProgress');
        expect(updateDashboardResponse.payload.dashboard).toMatchObject({
            ...pick(dashboard, 'id', 'name', 'kind', 'status'),
            hasDraft: true,
            permission: 'Owner',
        });

        subscription.unsubscribe();
    });

    it('receives update after permission change', async () => {
        // Step 1. Grant Owner permissions to the current dashboard to `user2`
        const permission = {
            id,
            type: 'UpdateDashboardPermissions',
            permissions: [{ user: otherClient.username, permission: 'Owner' }],
        };
        const permissionResponse = await client.request(permission);
        expect(permissionResponse.state).toBe('Done');

        // Step 2. Subscribe to current dashboard as `user2`
        const subscriber = vi.fn();
        const subscription = otherClient
            .send({ type: 'SubscribeToDashboard', id })
            .subscribe(subscriber);

        // Step 3. Change permissions to current dashboard to `Viewer` for `user2`
        const permissionsResponse = await client.request({
            ...permission,
            permissions: [{ user: otherClient.username, permission: 'Viewer' }],
        });
        expect(permissionsResponse.state).toBe('Done');
        expect(permissionsResponse.payload.type).toBe('PermissionsUpdated');

        // Step 4. Check that `user2` has received updated `Viewer` permission
        await wait();
        const { length } = subscriber.mock.calls;
        expect(length).toBeGreaterThanOrEqual(2);
        const lastResponse = subscriber.mock.calls[length - 1][0];
        expect(lastResponse.state).toBe('InProgress');
        expect(lastResponse.payload.dashboard).toMatchObject({
            id,
            permission: 'Viewer',
            ...pick(initialDashboard, 'name', 'kind', 'status'),
        });

        subscription.unsubscribe();
    });

    it('receives error after permission change to `None`', async () => {
        // Step 1. Grant Owner permissions to the current dashboard to `user2`
        const permission = {
            id,
            type: 'UpdateDashboardPermissions',
            permissions: [{ user: otherClient.username, permission: 'Owner' }],
        };
        const permissionResponse = await client.request(permission);
        expect(permissionResponse.state).toBe('Done');

        const subscriber = vi.fn();
        const subscription = otherClient
            .send({ type: 'SubscribeToDashboard', id })
            .subscribe(subscriber);

        // Step 3. Change permissions to current dashboard to `None` for `user2`
        const permissionsResponse = await client.request({
            ...permission,
            permissions: [{ user: otherClient.username, permission: 'None' }],
        });
        expect(permissionsResponse.state).toBe('Done');
        expect(permissionsResponse.payload.type).toBe('PermissionsUpdated');

        // Step 4. Check that `user2` has received subscription error
        await wait();
        expect(subscriber.mock.calls).toHaveLength(2);
        const secondResponse = subscriber.mock.calls[1][0];
        expect(secondResponse.error).toMatchSnapshot();

        subscription.unsubscribe();
    });

    it('unsubscribes from subscription', async () => {
        // Step 1. Subscribe to a dashboard.
        const subscriber = vi.fn();
        const correlationId = generateCorrelationId();
        const subscription$ = client
            .send({ type: 'SubscribeToDashboard', id }, { correlationId })
            .pipe(share());

        const firstResponse = await firstValueFrom(subscription$);
        expect(firstResponse.state).toBe('InProgress');
        // @ts-ignore TODO: remove this ts-ignore after tests migration to grpc
        const { dashboard } = firstResponse.payload as { dashboard: TDashboardItem };

        const subscription = subscription$.subscribe(subscriber);

        // Step 2. Unsubscribe.
        const unsubscribeResponse = await client.request(
            { type: 'Unsubscribe' },
            { correlationId },
        );
        expect(unsubscribeResponse.state).toBe('Done');
        expect(unsubscribeResponse.payload).toMatchSnapshot();

        // Step 3. Update dashboard
        const updatedDashboard = {
            type: 'UpdateDashboard',
            ...pick(dashboard, 'id', 'name', 'status', 'digest'),
            config: 'modified config',
        };
        const updateResponse = await client.request(updatedDashboard);
        expect(updateResponse.state).toBe('Done');

        // Step 4. Wait for another couple of seconds for any late updates
        // and check that we didn't receive any updates at all.
        await wait();
        expect(subscriber.mock.calls).toHaveLength(0);

        subscription.unsubscribe();
    });

    it('receives update after sharing', async () => {
        // Step 1. Subscribe to a dashboard.
        const subscriber = vi.fn();
        const correlationId = generateCorrelationId();
        const subscription$ = client
            .send({ type: 'SubscribeToDashboard', id }, { correlationId })
            .pipe(share());
        const subscription = subscription$.subscribe(subscriber);

        // Step 2. Share dashboard
        const shareResponse = await client.request({
            type: 'UpdateDashboardShareSettings',
            id,
            sharePermission: SharePermission.Viewer,
        });
        expect(shareResponse.state).toBe('Done');

        // Step 3. Wait for another couple of seconds for any late updates
        await wait();

        // Step 4. Check that update has been received
        expect(subscriber.mock.calls).toHaveLength(2);
        const initialDashboardResponse = subscriber.mock.calls[0][0];
        const updateDashboardResponse = subscriber.mock.calls[1][0];
        expect(initialDashboardResponse.payload.dashboard.sharePermission).toBe(
            SharePermission.None,
        );
        expect(updateDashboardResponse.payload.dashboard.sharePermission).toBe(
            SharePermission.Viewer,
        );

        subscription.unsubscribe();
    });

    it('receives updatable list of owners', async () => {
        // Step 1. Subscribe to a dashboard.
        const subscriber = vi.fn();
        const correlationId = generateCorrelationId();
        const subscription$ = client
            .send({ type: 'SubscribeToDashboard', id }, { correlationId })
            .pipe(share());
        const subscription = subscription$.subscribe(subscriber);

        // Step 2. Add another owner
        const permission = {
            id,
            type: 'UpdateDashboardPermissions',
            permissions: [{ user: otherClient.username, permission: 'Owner' }],
        };
        await client.request(permission);

        // Step 3. Wait for another couple of seconds for any late updates
        await wait();

        // Step 4. Check that update has been received
        const { length } = subscriber.mock.calls;
        expect(length).toBeGreaterThanOrEqual(2);
        expect(subscriber.mock.calls[0][0].payload.dashboard.owners).toEqual([
            { user: client.username },
        ]);
        expect(subscriber.mock.calls[length - 1][0].payload.dashboard.owners).toEqual([
            { user: client.username },
            { user: otherClient.username },
        ]);

        subscription.unsubscribe();
    });

    it('receives updatable permissions count as owner', async () => {
        // Step 1. Subscribe to a dashboard.
        const subscriber = vi.fn();
        const correlationId = generateCorrelationId();
        const subscription$ = client
            .send({ type: 'SubscribeToDashboard', id }, { correlationId })
            .pipe(share());
        const subscription = subscription$.subscribe(subscriber);

        // Step 2. Add another users to permissions list
        const permission = {
            id,
            type: 'UpdateDashboardPermissions',
            permissions: [{ user: otherClient.username, permission: 'Viewer' }],
        };
        await client.request(permission);

        // Step 3. Wait for another couple of seconds for any late updates
        await wait();

        // Step 4. Check that update has been received
        expect(subscriber.mock.calls).toHaveLength(2);
        expect(subscriber.mock.calls[0][0].payload.dashboard.permissionsCount).toBe(0);
        expect(subscriber.mock.calls[1][0].payload.dashboard.permissionsCount).toBe(1);

        subscription.unsubscribe();
    });

    it('does not receive permissions count as non-owner', async () => {
        // Step 1. Add another user as Editor
        const permission = {
            id,
            type: 'UpdateDashboardPermissions',
            permissions: [{ user: otherClient.username, permission: 'Editor' }],
        };
        await client.request(permission);

        // Subscribe to dashboard as Viewer user
        const response = await otherClient.request({ type: 'SubscribeToDashboard', id });

        // Step 4. Check that permissionsCount doesn't exist in the response
        expect(response.payload.dashboard.permissionsCount).toBeUndefined();
    });
});
