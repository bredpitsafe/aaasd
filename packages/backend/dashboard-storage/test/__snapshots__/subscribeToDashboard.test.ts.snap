// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SubscribeToDashboard > gets error when subscribing to non-existent dashboard 1`] = `
{
  "description": "Failed to subscribe to dashboard: You do not have permission to view this dashboard",
  "kind": "Authorization",
}
`;

exports[`SubscribeToDashboard > gets error when subscribing to other user\`s dashboard 1`] = `
{
  "description": "Failed to subscribe to dashboard: You do not have permission to view this dashboard",
  "kind": "Authorization",
}
`;

exports[`SubscribeToDashboard > receives error after permission change to \`None\` 1`] = `
{
  "description": "Failed to subscribe to dashboard: You do not have permission to view this dashboard",
  "kind": "Authorization",
}
`;

exports[`SubscribeToDashboard > unsubscribes from subscription 1`] = `
{
  "type": "Unsubscribed",
}
`;
